systest/test_rsyslog.py

```diff
@@ -35,14 +36,16 @@ def __init__(self, *, workdir, logfile, port):
 
         self.port = port
         self.conffile = "{}/rsyslogd.conf".format(workdir)
-        self.process = None
+        self.process: Optional[Popen] = None
 
         with open(self.conffile, "w") as fp:
             print(RSYSLOGD_TCP_CONF.format(logfile=logfile, port=port), file=fp)
 
     def _wait_until_running(self):
         # Wait until the rsyslogd port is available, but if it is not up in
         # five seconds assume that it has failed to start
+        if self.process is None:
+            raise RuntimeError("start must be called before invoking this function")
```

Comment:
can't we just call start instead of failing? I know it's not the intent of the api, but this seems like a hard fail that could be avoided

Reply:
This is an internal function that will literally never be called anywhere else except the function right below it. This check does only 2 things: makes mypy realize self.process is not None (without changing entirely how this particular API works right now), and gives a great big warning to any future test programmer that they've messed up (but really mostly the former).

Reasonably a new PR would just move the content of `_wait_until_running` into `start` which resolves all the ambiguity entirely...but then really, this should actually be a pytest fixture which uses `yield` semantics. The issue is all of that is varying levels of surgery which doesn't get closer to having rigorous type-checks over the code.